Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::SegmentedVector

cdumez opened the pull request:
#### 91f09395a0314120ee71a4c6651ad7a0b7422a7f
<pre>
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::SegmentedVector
<a href="https://bugs.webkit.org/show_bug.cgi?id=295874">https://bugs.webkit.org/show_bug.cgi?id=295874</a>

Reviewed by Darin Adler.

This tested as performance neutral on Speedometer and JetStream.

* Source/WTF/wtf/SegmentedVector.h:

Canonical link: <a href="https://commits.webkit.org/297349@main">https://commits.webkit.org/297349@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b41736b706d9aa57266ff94c222ecde2399c435f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111459 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31125 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21590 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117491 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61727 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31806 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39707 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/84714 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114406 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25398 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100334 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65162 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24746 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18469 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61313 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/103952 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94785 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18535 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120714 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/110013 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38508 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28620 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/93638 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38884 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96602 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93464 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23805 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38571 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16342 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34545 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38397 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43874 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/134288 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38062 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36192 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41395 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39764 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/SegmentedVector.h:
```diff
@@ -222,14 +221,14 @@ namespace WTF {
             T entries[0];
+
+            std::span<T, SegmentSize> entriesSpan() { return unsafeMakeSpan<T, SegmentSize>(entries, SegmentSize); }
         };
```
Seems like it would be better if the `entries` array was private. We should always access it through the span. Maybe even name them `m_entriesArray` and `entries`.

darinadler commented on Source/WTF/wtf/SegmentedVector.h:
```diff
@@ -222,14 +221,14 @@ namespace WTF {
-        void deleteAllSegments()
+        void destroyAllItems()
         {
             for (size_t i = 0; i < m_size; ++i)
                 at(i).~T();
```
Might be worth investigating whether we can use `std::ranges::destroy` here at some point.

darinadler commented on Source/WTF/wtf/SegmentedVector.h:
```diff
@@ -265,15 +264,13 @@ namespace WTF {
 
         size_t m_size { 0 };
-        Vector<Segment*, 0, CrashOnOverflow, 16, Malloc> m_segments;
+        Vector<MallocPtr<Segment>, 0, CrashOnOverflow, 16, Malloc> m_segments;
```
Can we find a way to use something more like `std::unique_ptr<Segment[]>` instead of `MallocPtr`? I ask because `MallocPtr.h` says “do not use this in new code”.

darinadler commented on Source/WTF/wtf/SegmentedVector.h:
```diff
@@ -29,11 +29,10 @@
 
 #include <wtf/MallocCommon.h>
+#include <wtf/MallocPtr.h>
 #include <wtf/Noncopyable.h>
```
Pretty sure we don’t need to include `MallocCommon.h` any more and we definitely don’t need `Noncopyable.h` since `MallocPtr.h` already includes that.
```suggestion
#include <wtf/MallocPtr.h>
```

darinadler reviewed (approved):
Seems great as is. A few questions that might lead to future improvements.

cdumez committed:
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::SegmentedVector
https://bugs.webkit.org/show_bug.cgi?id=295874

Reviewed by Darin Adler.

This tested as performance neutral on Speedometer and JetStream.

* Source/WTF/wtf/SegmentedVector.h:

Canonical link: https://commits.webkit.org/297349@main
